### PR TITLE
Fix variable declaration in default argument block

### LINKED ISSUE #11695: Potential regression with pattern matching inside default argument expressions
Hello.

I’ve come across a difference in behaviour between Elixir 1.9.4 and later Elixir versions (tested with 1.10, 1.10.4, 1.11.0, 1.11.2, 1.12, 1.12.3).

I'm not sure if it's an expected change, or an unexpected regression, but I couldn't find anything related in the changelogs.

I also posted to the forum so apologies for the duplication (wasn't sure if this was more likely to reach the correct people): https://elixirforum.com/t/potential-regression-with-pattern-matching-inside-default-argument-expressions/46473

### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
* Elixir: Expected behaviour on 1.9.4, regression on versions 1.10, 1.10.4, 1.11.0, 1.11.2, 1.12, 1.12.3.
* Erlang: Erlang/OTP 23 [erts-11.2.2.9] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]
* Operating system: MacOS 12.2.1

### Reproduction steps

Here is a test module showing the issue:

```elixir
defmodule TestDefaultArguments do
  def test(
        arg1,
        arg2 \\ (
          default = "default"
          default
        )
      ) do
    arg1 <> "-" <> arg2
  end
end
```

If you run then `mix run -e "TestDefaultArguments.test(\"arg1\") |> IO.puts()"`:

### Expected behaviour

On Elixir 1.9.4, you see the expected output `arg1-default`.

### Actual behaviour

On later Elixir versions (tested on Elixir 1.10, 1.10.4, 1.11.0, 1.11.2, 1.12, 1.12.3), you see the following error message:

```
** (MatchError) no match of right hand side value: "default"
    (test_gettext 0.1.0) lib/test_gettext.ex:5: TestDefaultArguments.test/1
    (stdlib 3.14.2.2) erl_eval.erl:680: :erl_eval.do_apply/6
    (stdlib 3.14.2.2) erl_eval.erl:888: :erl_eval.expr_list/6
    (stdlib 3.14.2.2) erl_eval.erl:411: :erl_eval.expr/5
    (elixir 1.10.4) lib/code.ex:341: Code.eval_string_with_error_handling/3
    (elixir 1.10.4) lib/enum.ex:783: Enum."-each/2-lists^foreach/1-0-"/2
```

### Further information

I believe it’s something to do with the pattern match `default = "default"` within the default argument expression.

I came across this in a situation where the default argument expression is `gettext("foo")`. This is a macro which compiles down to:

```elixir
(
  msgid = ManagerWeb.Gettext.dpgettext_noop("default", nil, "foo")
  Gettext.dpgettext(ManagerWeb.Gettext, "default", nil, msgid, %{})
)
```

And I get a similar error message:

```
** (MatchError) no match of right hand side value: "foo"
```

So what I care about doing is having the gettext call in the default argument expression, but I’ve tried to distill it down to what I think is going wrong.

Thanks for your help!

### PR DESCRIPTION
**Pull request by sabiwara**
Hi,

I think this fix should solve this issue:

Close #11695

If my understanding is correct, there was a `version` conflict between `arg1` and `arg2`.
`:elixir_def.expand_defaults` would emit `version: 0` for the default arg2, but the version counter would not be passed and start at 0 again for `unpack_expanded` which would emit `version: 0` for arg1 as well.

Regarding tests: I couldn't find a relevant test file to add this case, so I added a new file, please let me know if there is a better place.

### CODE REVIEW

File: lib/elixir/test/elixir/kernel/def_test.exs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+Code.require_file("../test_helper.exs", __DIR__)
+
+defmodule Kernel.DefTest do
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/def_test.exs:3]:
```suggestion
defmodule Kernel.DefaultsTest do
```

Can you also please rename the test file accordingly?

File: lib/elixir/test/elixir/kernel/def_test.exs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        ),
+        z \\ (
+          default = "z"
+          default
+        )
+      ) do
+    {x, y, z}
+  end
+
+  test "function with block defaults" do
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/def_test.exs:20]:
```suggestion
  test "with block defaults" do
```

File: lib/elixir/test/elixir/kernel/def_test.exs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    assert {1, "y", "z"} = fun_with_block_defaults(1)
+    assert {1, 2, "z"} = fun_with_block_defaults(1, 2)
+    assert {1, 2, 3} = fun_with_block_defaults(1, 2, 3)
+  end
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/def_test.exs:24]: Let's also add a test that accessing a variable from a default block raises. Something like this:

```
def test(_ \\ (default = 1)), do: default
```

You can use something like:

assert_raise ..., fn ->
  defmodule VarDefaultScope do
    def test(_ \\ (default = 1)), do: default
  end
end
```

**Comment by josevalim** [comment]
Awesome job, i have added some comments.

File: lib/elixir/test/elixir/kernel/def_test.exs (hunk 3 above)
Comment (sabiwara) [reply lib/elixir/test/elixir/kernel/def_test.exs]: Added in [3cdb863](https://github.com/elixir-lang/elixir/pull/11758/commits/3cdb863f35cdab1ce368a5bdfb1f53ac057eee55)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**